### DISCUSSION TITLE
Schedules are not executed. APScheduler 4.0.0a4

### DISCUSSION THREAD

**Discussion by klaydev** [agronholm/apscheduler#821] [Q&A]
I have this code:
```py
data_store = SQLAlchemyDataStore(engine)
event_broker = AsyncpgEventBroker.from_async_sqla_engine(engine)
scheduler_maker = AsyncScheduler(data_store, event_broker)

async def op():
    logging.info("op")

async def some_func():
    async with scheduler_maker as scheduler:
        await scheduler.add_schedule(op, IntervalTrigger(seconds=5))

async def main():
        async with schedules.scheduler_maker as scheduler:
             await scheduler.start_in_background()
asyncio.run(main())
```
In the logs i have:
```py
INFO:apscheduler._schedulers.async_:Connection established
INFO:apscheduler._schedulers.async_:Added new schedule (task=Task(id='src.modules.manufactorum.schedule:op', func='src.modules.manufactorum.schedule:op', job_executor='async', max_running_jobs=None, misfire_grace_time=None), trigger=IntervalTrigger(seconds=5, start_time='2023-11-23 23:13:37.373670+03:00')); next run time at 2023-11-23 23:13:37.373670+03:00
INFO:apscheduler._schedulers.async_:Stream finished
```
PS: some_func calls in other place
And nothing...
I don't understand why schedules don't work, because everything is done according to the instructions from the documentation.

**Comment by agronholm** [comment]
In `main()`, you're starting the scheduler and immediately stopping it (since the context manager block is exited immediately). How did you expect this to work?

**Reply by agronholm** [reply]
Also, you should only start the scheduler in one place. It's not clear where `some_func()` fits into this picture.

**Reply by agronholm** [reply]
Perhaps `await scheduler.run_until_stopped()` is what you were after?

**Reply by klaydev** [reply]
> In `main()`, you're starting the scheduler and immediately stopping it (since the context manager block is exited immediately). How did you expect this to work?

I expected that after launching the context manager, the scheduler would continue its execution. yes, because of the definition that I read start in the background. it is executed in a separate thread and as I understand it, this thread should remain in principle throughout the existence of the scheduler.

**Reply by klaydev** [reply]
> Also, you should only start the scheduler in one place. It's not clear where `some_func()` fits into this picture.

I tried to run. scheduler. in the main function, the version of the code that I threw off to you is my last attempt to do anything, as I even run the main function scheduler, it stopped working.

**Reply by klaydev** [reply]
> Perhaps `await scheduler.run_until_stopped()` is what you were after?

unfortunately, this does not suit me, since this function blocks the execution of the entire program, and I need parallel execution of the scheduler and the rest of the program.

**Reply by klaydev** [reply]
as I understand it, the scheduler must always be in a running state to process all the tasks that come to it at all, since it is not always known when the task will come to start the scheduler by this time, it also does not always work out in my understanding, or to act on the basis of an infinite loop, but it seems to me that this approach is not entirely logical

**Reply by agronholm** [reply]
As it says, `start_in_background()` does start the scheduler in the background, but then that call returns, and then it exits the context manager which causes the scheduler to be shut down. This is why there are two very different methods for starting the scheduler: one runs it in a background task while the other keeps running it in the current task until the scheduler is stopped or the task is cancelled. These examples in the documentation are side by side. Did you miss the foreground example there?

**Reply by agronholm** [reply]
> as I understand it, the scheduler must always be in a running state to process all the tasks that come to it at all

Correct.

> since it is not always known when the task will come to start the scheduler by this time

Can you rephrase this? I don't understand what you mean here.

> or to act on the basis of an infinite loop

Meaning?

**Reply by agronholm** [reply]
It's hard to suggest the proper solution when I don't know the first thing about your app. But the gist here is that you need to have one task dedicated to running the scheduler. It doesn't have to be the main task.

**Reply by klaydev** [reply]
> Can you rephrase this? I don't understand what you mean here.


I mean, in my case, the time when tasks arrive is not always known, and accordingly, it is not always known when to start the scheduler.

**Reply by klaydev** [reply]
I mean, the scheduler can be run in a conditional other thread, in an infinite loop. but I don't see much logic in this action.

**Reply by agronholm** [reply]
How about starting it when your application starts?

**Reply by klaydev** [reply]
perhaps there is some other way to run the scheduler not through an infinite loop, but this does not occur to me.

**Reply by klaydev** [reply]
> How about starting it when your application starts?

I tried to do this, but it didn't have any effect.

**Reply by agronholm** [reply]
You need to tell me more about your app. How did you try to start the scheduler?

**Reply by klaydev** [reply]
> You need to tell me more about your app. How did you try to start the scheduler?

my application is a Discord bot. tasks are added after calling some Discord command that requires some actions after a certain time (some_func is an example).
I tried to run the scheduler in different ways. in the main function, using both methods, directly in some_func itself, I created a separate thread and started it there, but nothing helped

**Reply by agronholm** [reply]
> my application is a Discord bot

I hope you're not assuming that it's obvious to me what library one uses to make async Discord bots.
How is that app started then? Show me some code please.

> I created a separate thread and started it there, but nothing helped

By thread you mean task? How did that code look like?

**Reply by klaydev** [reply]
Sorry, I'm not at the PC right now. I'll send it tomorrow

**Reply by klaydev** [reply]
Let me form the question differently. Is it possible to leave the scheduler always running? Because I have some tasks that require a DateTrigger as a trigger. Sample code below
```py
# some_module.py
data_store = SQLAlchemyDataStore(engine)
event_broker = AsyncpgEventBroker.from_async_sqla_engine(engine)
scheduler_maker = AsyncScheduler(data_store, event_broker)

async def add_schedule(some_arg):
        # Some logic here
        async with scheduler_maker as scheduler:
             await scheduler.add_schedule(some_func, DateTrigger(run_date), args=(some_arg,))

async def some_func(some_arg):
     await add_schedule(some_arg)
     # Some logic here
```
```py
# another_module.py
from some_module import add_schedule

async def some_command_handler(*args, **kwargs):
    # Some logic here
    await add_schedule(something)
```
And the thing is that some_command_handler can be called at any time, i.e. the time when the scheduler needs to be started is unknown in advance. In addition, Datatrigger is used, which imposes some difficulties with the time of determining the start of the scheduler, because it is not entirely rational to run it every second, for example. In my opinion, the scheduler should run itself and perform the task in the case of DateTrigger. Correct me if I'm wrong

**Reply by agronholm** [reply]
> Let me form the question differently. Is it possible to leave the scheduler always running?

That is what you normally want to do. I'm not sure how else you expected to be able to make the scheduler work. You want to start the scheduler when your app starts and leave it running until the application is shut down. If you could just shed light on how you start your app in the first place, I could possibly give you some hints on how to manage the scheduler's life cycle.

**Reply by klaydev** [reply]
The application is launched via `bot.run(token)` https://discordpy.readthedocs.io/en/stable/ext/commands/api.html?highlight=run#discord.ext.commands.Bot.run

**Reply by agronholm** [reply]
Ok, so looks like you want to call `bot.start()` instead, like so:
```python
import asyncio

async def main() -> None:
    data_store = SQLAlchemyDataStore(engine)
    event_broker = AsyncpgEventBroker.from_async_sqla_engine(engine)
    async with AsyncScheduler(data_store, event_broker) as scheduler:
        await scheduler.start_in_background()
        await bot.start()

asyncio.run(main())
```

**Reply by klaydev** [reply]
Unfortunately, it won't work that way, because `scheduler.start_in_background` creates an event loop, and `boot.start` also creates an event loop, resulting in an Exception

**Reply by agronholm** [reply]
No, `AsyncScheduler.start_in_background()` doesn't start an event loop. I should know, I wrote the code. Maybe you're confusing that with the method with the same name in the synchronous scheduler?

**Reply by agronholm** [reply]
And regarding `bot.start()`, where in the Discord documentation does it say that it creates an event loop?

**Reply by agronholm** [reply]
You got it to work?

**Reply by agronholm** [reply]
Could you mark this as the accepted answer if you got it to work?
